db/commitlog/commitlog.cc

```diff
@@ -1379,18 +1439,69 @@ future<> db::commitlog::segment_manager::message_loop() {
         // TODO: ordering of which we issue first.
 
         if (!messages.empty()) {
+            tasks = do_writes(std::move(tasks), messages);
```

Comment:
This mixing of free futures and coroutines is confusing.

I think there's also a problem - you only wait for tasks when existing the loop, so if you're adding tasks faster than you can retire them, you'll run out of memory.

Reply:
There is a risk we overload pending tasks, yes. But otoh, we will not be doing much more than we did in earlier iterations of commitlog, thus the amount of actual living tasks in the background should be limited. 
The task-queue-via-coroutine frames might be an issue. Any tasks in the chain will/might finish at some point while we are still in the message loop, and this should release his resources, and promote the result into the next (waiting) frame in the chain, which should execute in the background regardless if we co_await or not. Thus in the best case scenario, "tasks" should be a ready future more or less mostly when we wake up. And the resources inbetween released. Or is my brain missing something here?

Reply:
There's no real difference between a task due to a continuation and a task due to a coroutine (coroutines can take less memory since each continuation can be an allocation). The question is if we can add tasks faster than they are retired, and if there is a way for a reader of the code to see it.

Reply:
There is: I put in counters (not exposed, but readable in debugger) for pending continuations. 
And again: we will not add more tasks (continuations) to this chain than already exists in CL code today (but unbounded). 
The basic size of the set of background jobs taking up continuations in reactor queue and here will remain more or less the same. So that we don't really put in any bounds check or whatnot here is premeditated on this fact.